Tasks/Common/nuget-task-common/NuGetToolGetter.ts

```diff
@@ -20,65 +20,68 @@ enum NuGetReleaseStage
     "ReleasedAndBlessed"
 }
 
-export async function getNuGet(versionSpec: string): Promise<string> {
+const NUGET_TOOL_NAME: string = 'NuGet';
+const NUGET_EXE_FILENAME: string = 'nuget.exe';
+
+export async function getNuGet(versionSpec: string, checkLatest: boolean): Promise<string> {
     if (toolLib.isExplicitVersion(versionSpec)) {
+        // Check latest doesn't make sense when explicit version
+        checkLatest = false; 
         taskLib.debug('Exact match expected on version: ' + versionSpec);
     }
     else {
         taskLib.debug('Query match expected on version: ' + versionSpec);
         console.log(taskLib.loc("Info_ExpectBehaviorChangeWhenUsingVersionQuery"));
     }
 
-    let localVersions: string[] = toolLib.findLocalToolVersions('NuGet');
+    // If we're not checking latest, check the cache first
+    let toolPath: string;
+    if (!checkLatest) {
+        taskLib.debug('Trying to get tool from local cache');
+        toolPath = toolLib.findLocalTool(NUGET_TOOL_NAME, versionSpec);
+    }
+
+    let localVersions: string[] = toolLib.findLocalToolVersions(NUGET_TOOL_NAME);
     let version: string = toolLib.evaluateVersions(localVersions, versionSpec);
-    taskLib.debug('Trying to get tool from local cache');
-    if (version) {
+
+    if (toolPath) {
+        // If here, then we're not checking latest and we found the tool in cache
         console.log(taskLib.loc("Info_ResolvedToolFromCache", version));
     }
     else {
-        taskLib.debug('Querying versions list');
-        
-        let versionsUrl = 'https://dist.nuget.org/tools.json';
-        let rest: restm.RestClient = new restm.RestClient('vsts-tasks/NuGetToolInstaller');
-        
-        let nugetVersions: INuGetVersionInfo[] = (await rest.get<INuGetVersionInfo[]>(versionsUrl, GetRestClientOptions())).result;
-        let availableVersions: INuGetVersionInfo[] = nugetVersions.filter(x => x.stage !== NuGetReleaseStage.EarlyAccessPreview);
-        let filteredVersions: string[] = availableVersions.map(x => x.version);
-        let versionUrlMap: Map<string, string> = new Map<string, string>();
-        availableVersions.forEach((versionInfo:INuGetVersionInfo) => {
-            versionUrlMap[versionInfo.version.toLowerCase()] = versionInfo.url;
-        });
-
-        version = toolLib.evaluateVersions(filteredVersions, versionSpec);
-        if(!version)
-        {
-            taskLib.error(taskLib.loc("Error_NoVersionWasFoundWhichMatches", versionSpec)); 
-            taskLib.error(taskLib.loc("Info_AvailableVersions"));
-            taskLib.error(availableVersions.map(x => x.version).join(";"));
-            throw new Error(taskLib.loc("Error_NuGetToolInstallerFailer", "NuGet"));
+        let versionInfo: INuGetVersionInfo = await getLatestMatchVersionInfo(versionSpec);
```

Comment:
I think this only needs to get called if checkLatest is true, or if we cannot find a local version, otherwise you're always checking for the latest even if the user doesn't want to.